Add robotest for create-read events, add retries config parameter

dadlex opened the pull request.

dadlex committed:
Correct robotest for create-read events and add missed js script
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Add missed js script and correct the robotest a bit

Signed-off-by: Alexander <alex15051997@yandex.com>

akurilov commented on src/test/robot/api/storage/pravega.robot:
```diff
@@ -47,6 +50,18 @@ Create Byte Streams Test
     Log  ${std_out}
     Validate Metrics Total Log File  ${step_id}  CREATE  ${count_limit}  0  104857600
 
+Create and read Events Test
```
please rename to Read Events Test

akurilov replied:
Please add some concurrency, 10 for example

dadlex replied:
done

akurilov commented on src/test/robot/api/storage/data/create_and_read_events.js:
```diff
@@ -0,0 +1,45 @@
+            "namespace" : "goose"
+        }
+	})
+	.run();
```
missing newline

dadlex replied:
done

akurilov commented on src/test/robot/api/storage/data/create_and_read_events.js:
```diff
@@ -0,0 +1,45 @@
+            	}
+        	}
+    	},
+        "storage" : {
```
move to the shared config or command line arguments (better)

dadlex replied:
done

akurilov commented on src/test/robot/api/storage/data/create_and_read_events.js:
```diff
@@ -0,0 +1,45 @@
+PreconditionLoad
+	.config({
+		"item" : {
+			"type" : "data",
```
not needed, by default

dadlex replied:
fixed

akurilov commented on src/test/robot/api/storage/data/create_and_read_events.js:
```diff
@@ -0,0 +1,45 @@
+ReadLoad
+	.config({
+	    "item" : {
+	        "type" : "data",
```
not needed, by default

dadlex replied:
fixed

akurilov commented on src/test/robot/api/storage/data/create_and_read_events.js:
```diff
@@ -0,0 +1,45 @@
+	            "path" : "test"
+	        },
+            "data" : {
+                "size" : "1000KB"
```
not needed

dadlex replied:
fixed

akurilov commented on src/test/robot/api/storage/data/create_and_read_events.js:
```diff
@@ -0,0 +1,45 @@
+		"item" : {
+			"type" : "data",
+			"output" : {
+				"path" : "test"
```
use single value specified by the environment variable for both write and read
see other fun tests for example

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -743,30 +760,58 @@ void submitEventReadOperation(final O evtOp, final String nodeAddr) {
-			val readerGroupManager = readerGroupManagerCache.computeIfAbsent(scopeName, readerGroupManagerCreateFunc);
-			readerGroupManager.createReaderGroup(readerGroup, readerGroupConfig);
+			evtReaderGroupManagerCache.computeIfAbsent(
+				scopeName, key -> {
```
use a method reference instead of creating a new lambda every time

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -743,30 +760,58 @@ void submitEventReadOperation(final O evtOp, final String nodeAddr) {
-			val readerGroupManagerCreateFunc = readerGroupManagerCreateFuncCache.computeIfAbsent(
+			val readerGroupConfigBuilder = evtReaderGroupConfigBuilder.get();
+			val readerGroupConfig = evtReaderGroupConfigCache.computeIfAbsent(
+				scopeName + SLASH + streamName, key -> readerGroupConfigBuilder.stream(key).build()
```
use a method reference instead of creating a new lambda every time

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -383,11 +396,12 @@ protected String requestNewAuthToken(final Credential credential) {
 		} else {
-			items = makeSingleEventItem(itemFactory, path, prefix, lastPrevItem);
+			//+1 is needed to correctly work with the end of the stream
+			items = makeEventItems(itemFactory, path, prefix, lastPrevItem, retries+1);
```
spaces around "+"

dadlex replied:
done

dadlex committed:
Add docs on the new config parameter, add new config parameter to the integration tests
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Add missed js script and correct the robotest a bit

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Add docs on the new config parameter, add new config parameter to the integration tests
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Add missed js script and correct the robotest a bit

Signed-off-by: Alexander <alex15051997@yandex.com>

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -743,30 +760,58 @@ void submitEventReadOperation(final O evtOp, final String nodeAddr) {
+							completeOperation(evtOp, FAIL_TIMEOUT);
+						} else {
+							lastFailedStreamPos = streamPos;
+							Loggers.ERR.warn("{}: corrupted event @ position {}", stepId, streamPos);
```
return false, doesn't invoke `completeOperation`

dadlex replied:
done

akurilov commented on README.md:
```diff
@@ -227,11 +227,15 @@ docker run \
 > * Works synchronously
 
-There is also another option, called `storage-driver-read-timeoutMillis`. Pravega documentation says it only works when
+There is a specific option (config parameter) for reading called `storage-driver-event-timeoutMillis`. Pravega documentation says it only works when
```
recycle mode notes, example

dadlex replied:
done

akurilov replied:
example command to run the events reading is also necessary

akurilov commented on src/test/robot/api/storage/data/read_events.js:
```diff
@@ -0,0 +1,39 @@
+	.config({
+		"item" : {
+            "data" : {
+                "size" : "1000KB"
```
the bug related to the 1MB events write was fixed in the current pravega version already, so this option may be skipped

dadlex replied:
done

akurilov commented on src/test/java/com/emc/mongoose/storage/driver/pravega/integration/ReadEventsTest.java:
```diff
@@ -0,0 +1,108 @@
+import static com.emc.mongoose.base.item.op.Operation.SLASH;
+import static com.emc.mongoose.base.item.op.Operation.Status.SUCC;
+import static org.junit.Assert.assertEquals;
+
```
add the `@Ignore` annotation, will fail the test otherwise

dadlex replied:
done

akurilov commented on src/main/resources/config-schema-storage-pravega.yaml:
```diff
@@ -9,6 +9,7 @@ storage:
         enabled: boolean
         count: long
       timeoutMillis: int
+      retries: int
```
not necessary anymore

dadlex replied:
done

dadlex committed:
Delete a config parameter "retries", correct js script for robotest, add Ignore annotation in ReadEventsTest
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Delete retries parameter, delete docs for it. Make some corrections on readEvents.js.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex commented:
Methods references and return value will be done in the next commit

dadlex committed:
Make submitEventRead method return false in case we re-read the failed event, delete retries param from CommonTest
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Make submitEventRead method return boolean, as well as submitEvent

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Add duration and latency into submitEventRead
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Duration is started right before readNextEvent() and finished right after. Latency is measured right before countBytesDone().
This description is obvious, but it points out what should be double-checked on review.

Signed-off-by: Alexander <alex15051997@yandex.com>

akurilov commented on README.md:
```diff
@@ -227,11 +227,21 @@ docker run \
+Learn more about recycle-mode here: https://github.com/emc-mongoose/mongoose-base/tree/master/doc/design/recycle_mode.
+
+There is one important fact considering the end of a stream. As we do not know the size of a stream,
+we only find out that it is the end, when `readNextEvent()` returns false, so 1 failed operation is not
```
`readNextEvent` returns the `EventRead` instance which returns `null` on `getEvent` call

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -502,6 +517,7 @@ protected boolean prepare(final O operation) {
 	@Override
 	protected final boolean submit(final O op)
 					throws IllegalStateException {
+		boolean eventOpHasFailed = false;
```
boolean ok = concurrencyThrottle.tryAcquire();

dadlex replied:
I do not have a better suggestion for now, but isn't "ok" a bit too vague?

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -510,7 +526,9 @@ protected final boolean submit(final O op)
 				final String nodeAddr = op.nodeAddr();
 				switch (streamDataType) {
 				case EVENTS:
-					submitEventOperation(op, nodeAddr);
```
ok = submitEventOperation(op, nodeAddr);

dadlex replied:
Slick

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -502,6 +517,7 @@ protected boolean prepare(final O operation) {
 	protected final boolean submit(final O op)
 					throws IllegalStateException {
+		boolean eventOpHasFailed = false;
 		if (concurrencyThrottle.tryAcquire()) {
```
if(ok) {

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -519,6 +537,9 @@ protected final boolean submit(final O op)
+				return false;
+			}
 			return true;
 		} else {
```
return ok;

dadlex replied:
I would prefer returning ok from inside the if statement, and explicitly return false from outside if. I hope it's the same idea here

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -552,18 +573,22 @@ void submitNoop(final O op) {
 			submitEventCreateOperation(op, nodeAddr);
 			break;
 		case READ:
-			submitEventReadOperation(op, nodeAddr);
```
ok = submitEventReadOperation(op, nodeAddr);

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -552,18 +573,22 @@ void submitNoop(final O op) {
-	void submitEventOperation(final O op, final String nodeAddr) {
+	boolean submitEventOperation(final O op, final String nodeAddr) {
 		val type = op.type();
+		boolean returnValue = true;
```
boolean ok = true;

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -552,18 +573,22 @@ void submitNoop(final O op) {
 		default:
 			throw new AssertionError("Unsupported event operation type: " + type);
 		}
+		return returnValue;
```
return ok;

akurilov commented on src/test/robot/api/storage/pravega.robot:
```diff
@@ -92,19 +95,33 @@ Batch Create Event Stream Test
+Read Events Test
+    ${step_id} =  Set Variable  read_events_test
+    Remove Directory  ${LOG_DIR}/${step_id}  recursive=True
+    ${args} =  Catenate  SEPARATOR= \\\n\t
```
add the command line option
`...  --storage-net-node-addrs=${node_addr}`
the test will hang w/o it

dadlex replied:
Thanks for noticing. Repeating your mistakes

akurilov commented on src/test/robot/api/storage/data/read_events.js:
```diff
@@ -0,0 +1,34 @@
+
+PreconditionLoad
+	.config(sharedConfig)
+	.config({
```
`.config({})`
specify the count limit from the command line
this makes the scenario more independent and reusable

dadlex replied:
Agree, done.

akurilov replied:
where?

akurilov commented on src/test/robot/api/storage/pravega.robot:
```diff
@@ -92,19 +95,33 @@ Batch Create Event Stream Test
+    &{env_params} =  Create Dictionary SCOPE_NAME=scope6
+    ${std_out} =  Execute Mongoose Scenario  ${DATA_DIR}  ${env_params}  ${args}
+    Log  ${std_out}
+    Validate Metrics Total Log File  ${step_id}  READ  1000  2  1024000000
```
use variable for the count limit

dadlex replied:
done

akurilov commented on src/test/robot/api/storage/pravega.robot:
```diff
@@ -92,19 +95,33 @@ Batch Create Event Stream Test
+    ${step_id} =  Set Variable  read_events_test
+    Remove Directory  ${LOG_DIR}/${step_id}  recursive=True
+    ${args} =  Catenate  SEPARATOR= \\\n\t
+    ...  --load-step-id=${step_id}
```
add the command line option
`...  --load-op-limit-count=${count_limit}`

dadlex replied:
done

akurilov commented on ci/.gitlab-ci.yml:
```diff
@@ -150,11 +150,11 @@ test_read_all_byte_streams:
+  variables:
+    SUITE: api.storage
+    TEST: batch_create_event_stream
 
```
add your read events test into this file

dadlex committed:
Refactor submit returns for events, add robotest to gitlab ci, correct docs
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Change readme information on read events.
Refactor submitEventRead, submitEvent and submit a bit.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Change one lambda into function reference
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Add interface and implementation for readerGroupCache,
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Add new cache to reuse created readerGroups. Get rid of the second lambda.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Update Zlib version
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Update to 11.3 zlib

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Delete @Ignore in ReadEventsTest
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Insert back @Ignore in ReadEventsTest
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Correct whitespace in test_read_events robotest
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Merge branch 'master' into issue_PRAVEGA-84-byte_streams_append

dadlex committed:
Correct test_read_events robotest
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Merge remote-tracking branch 'origin/issue_PRAVEGA-84-byte_streams_append' into issue_PRAVEGA-84-byte_streams_append

dadlex committed:
Update README.md

dadlex committed:
Try
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Merge remote-tracking branch 'origin/issue_PRAVEGA-84-byte_streams_append' into issue_PRAVEGA-84-byte_streams_append

dadlex committed:
Try
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Try
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Correct read_events.js
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

load-op-limit-count should be specified for create only. Read should be working without limits as we always have N+1 operations (N events + end of the stream)

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Correct read_events.js
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

load-op-limit-count should be specified for create only. Read should be working without limits as we always have N+1 operations (N events + end of the stream)

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
try
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
try
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Alexander <alex15051997@yandex.com>
